### PR TITLE
New, more modern KivaFont trait

### LINKED ISSUE #417: Use new Trait types
This should perhaps wait for Enable 5.1 rather than 5.0, but we should use `Union` instead of `Either`, `Map` and other new trait types where appropriate.

If we can remove uses of `Trait` as well, that would be good.

### LINKED ISSUE #471: Clean-up Imports
These are directly lifted from https://github.com/enthought/enable/blob/master/TODO.txt

- [x] Clean-up the failing import of enthought.freetype in
  kiva.trait_defs.ui.wx.kiva_font_editor.
- [x] Clean-up imports of enable (I mean NOT enable2) done in:
    enable2/traits/ui/wx/enable_rgba_color_editor.py
- [ ] Clean-up direct imports of traitsui.wx

The first two I believe are no longer relevant (?)

### LINKED ISSUE #896: Integration with new Pyface Font class and Trait
At a minimum it would be nice to be able to specify fonts with Pyface Fonts - ideally we might be able to replace the Kiva fonts with Pyface fonts - this needs exploration.

### LINKED ISSUE #922: `KivaFontTrait` needs to delay import of `KivaFontEditor`
The current old-style `Trait` construction means that `KivaFontEditor` is imported at the top-level of the module.  This should be deferred by using `get_editor` method on a `TraitType` subclass or similar re-working of the editor.

This would be a good opportunity to remove the use of `Trait()` and similar old-style constructors.

Also, `KivaFontTrait` really belongs in Enable...

### PR DESCRIPTION
**Pull request by corranwebster**
This creates a KivaFont TraitType, which is:
- in enable, rather than Kiva, meaning Kiva can potentially not need traits
- has a better font parser that supports more weight types
- the font parser can be swapped out for something better easily
- uses the new KivaFontEditor, which is available for Qt
- converts Pyface Fonts to Kiva Fonts

This requires #928 to be merged before it is.

In conjunction with #924 this closes #417 
It also closes #922 and closes #896 and closes #471

### CODE REVIEW

**Comment by mdickinson** [comment]
I'll wait on review until #924 and #928 have been merged into main (and from there into this branch).

File: enable/trait_defs/kiva_font_trait.py
Comment (jwiggins) [inline enable/trait_defs/kiva_font_trait.py:97]: `FontParseError` is not defined in this scope

File: enable/trait_defs/kiva_font_trait.py
Comment (jwiggins) [inline enable/trait_defs/kiva_font_trait.py:80]: Should you mention `FontParseError` here as well?

File: kiva/trait_defs/kiva_font_trait.py
Comment (jwiggins) [inline kiva/trait_defs/kiva_font_trait.py:13]: Are you sure we want to import from `enable` here?

**Review by jwiggins — COMMENTED** [review]
Some preliminary comments

File: enable/tests/trait_defs/test_kiva_font_trait.py
Comment (corranwebster) [reply enable/tests/trait_defs/test_kiva_font_trait.py:82]: The different is that they are given a font object, rather then a descriptive string.  There _is_ an issue that `test_validate_pyface_font` should use a `PyfaceFont`.

File: kiva/trait_defs/kiva_font_trait.py
Comment (corranwebster) [reply kiva/trait_defs/kiva_font_trait.py:13]: Yes, I _think_ so.  It's preliinary to getting rid of `kiva.trait_defs`.

As long as we don't have circularity of imports it should be fine.

**Comment by corranwebster** [comment]
@mdickinson if you have time to review, this would be great.

**Comment by mdickinson** [comment]
Will review first thing tomorrow. Please could you fix the conflicts first?

File: enable/trait_defs/kiva_font_trait.py
Comment (mdickinson) [reply enable/trait_defs/kiva_font_trait.py:97]: I guess that also means that the test suite doesn't exercise this branch. (Either that, or the `NameError` is getting swallowed somewhere.) That may be worth fixing.

File: kiva/fonttools/font.py
Comment (mdickinson) [inline kiva/fonttools/font.py:90]:
```suggestion
    In addition, the parser ignores the terms 'pt', 'point', 'px', and 'family'.
```

File: kiva/fonttools/font.py
Comment (mdickinson) [inline kiva/fonttools/font.py:83]:
```suggestion
        The following decoration term is accepted: 'underline'
```

File: enable/trait_defs/ui/kiva_font_editor.py
Comment (mdickinson) [inline enable/trait_defs/ui/kiva_font_editor.py:26]: There are several unused imports in this file; not introduced in this PR, I know, but they may be worth cleaning up in passing.

File: enable/trait_defs/ui/kiva_font_editor.py
Comment (mdickinson) [inline enable/trait_defs/ui/kiva_font_editor.py:32]: Font question, possibly off-topic for this PR: I don't think I've met the term "Demi-bold" before. Is it completely interchangeable with semi-bold, and is there a reason to prefer the "Demi-bold" description over "Semi-bold"? The mix of the two terms feels confusing.

File: kiva/fonttools/font.py
Comment (mdickinson) [inline kiva/fonttools/font.py:126]: Nitpick: for case normalization, `str.casefold` is preferred over `str.lower` (it has better properties with respect to non-Roman character sets). It probably doesn't matter much in this instance.

File: kiva/fonttools/font.py
Comment (mdickinson) [inline kiva/fonttools/font.py:197]: `RuntimeError` seems like a strange choice here. `ValueError` would seem like a better fit.

File: kiva/fonttools/font.py
Comment (mdickinson) [inline kiva/fonttools/font.py:201]:
```suggestion
                f"Expected size to be an int, got {size!r}")
```

File: kiva/fonttools/font.py
Comment (mdickinson) [inline kiva/fonttools/font.py:204]:
```suggestion
                f"Expected family to be an int, got {family!r}")
```

File: kiva/fonttools/font.py
Comment (mdickinson) [inline kiva/fonttools/font.py:207]:
```suggestion
                f"Expected weight to be an int, got {weight!r}")
```

File: kiva/fonttools/font.py
Comment (mdickinson) [inline kiva/fonttools/font.py:210]:
```suggestion
                f"Expected style to be an int, got {style!r}")
```

File: kiva/fonttools/font.py
Comment (mdickinson) [inline kiva/fonttools/font.py:213]:
```suggestion
                f"Expected underline to be a int, got {underline!r}")
```

File: kiva/fonttools/font.py
Comment (mdickinson) [inline kiva/fonttools/font.py:216]:
```suggestion
                f"Expected encoding to be a bool, got {encoding!r}")
```

File: kiva/fonttools/font.py
Comment (mdickinson) [inline kiva/fonttools/font.py:196]: Please could we have some documentation of the meaning of the parameters? Some of them aren't obvious (like encoding).

For `encoding`: what does this represent? The validation checks that it's an `int`, the error message says it should be a `bool`, and the default value `DEFAULT` is taken from the list of families. I'm a bit confused.

File: kiva/fonttools/font.py
Comment (mdickinson) [inline kiva/fonttools/font.py:214]: Mismatch between the checked type (`int`) and the type in the error message (`bool`) here. I'm not sure which to believe: what's `encoding` for?

**Comment by corranwebster** [comment]
Looks like some imports got mangled/moved in the merge.  Not entirely sure how the tests are passing right now.

File: kiva/fonttools/font.py
Comment (mdickinson) [reply kiva/fonttools/font.py:216]: The line `raise RuntimeError("Bad value in Font() constructor.")` below this one is unreachable.

File: kiva/trait_defs/tests/test_kiva_font_trait.py
Comment (mdickinson) [inline kiva/trait_defs/tests/test_kiva_font_trait.py:20]: Suggest opening an issue for this.

**Review by mdickinson — COMMENTED** [review]
Done with first-pass review. Lots of comments, mostly at nitpick level.

At some point, it would be great to get flake8-style checks up and running on this repository; it's more efficient at catching things like `NameError`s than humans are.

File: enable/label.py
Comment (jwiggins) [reply enable/label.py:19]: If there is I doubt it's strictly adhered to

File: enable/tests/trait_defs/test_kiva_font_trait.py
Comment (jwiggins) [reply enable/tests/trait_defs/test_kiva_font_trait.py:34]: IIRC we cleaned up the code a year or two ago but never turned on flake8 as part of the CI process. I expect any new code to pass flake8.

File: enable/trait_defs/ui/kiva_font_editor.py
Comment (jwiggins) [reply enable/trait_defs/ui/kiva_font_editor.py:32]: https://developer.mozilla.org/en-US/docs/Web/CSS/font-weight#common_weight_name_mapping seems to support your preference

File: enable/label.py
Comment (corranwebster) [reply enable/label.py:1]: That's a sticky change - I fixed it in the previous PR and this one re-applied when I merged in main.  Very annoying.

```suggestion
# (C) Copyright 2005-2022 Enthought, Inc., Austin, TX
```

File: kiva/fonttools/font.py
Comment (corranwebster) [reply kiva/fonttools/font.py:126]: It may - fonts for non-Roman languages may have names that are not in Roman character sets.

There may be other places where this is an issue.

Comment (corranwebster) [reply kiva/fonttools/font.py:126]:
```suggestion
        lower_word = word.casefold()
```

File: kiva/fonttools/font.py
Comment (corranwebster) [reply kiva/fonttools/font.py:196]: I actually have no idea what `encoding` is used for.

File: kiva/fonttools/font.py
Comment (corranwebster) [reply kiva/fonttools/font.py:214]: Probably should be an int, given that was what was being checked for - `bool` is likely a copy-paste error in this case

File: kiva/fonttools/font.py
Comment (corranwebster) [reply kiva/fonttools/font.py:216]:
```suggestion
                f"Expected encoding to be n int, got {encoding!r}")
```

File: kiva/fonttools/font.py
Comment (corranwebster) [reply kiva/fonttools/font.py:196]: Also, I'm really not a fan of the error-checking that this class does - it's checking types, but a lot of the values are really supposed be restricted to a handful of constants.

It seems a bit pointless - likely better handled these days by static type checking or something.

File: kiva/fonttools/font.py
Comment (corranwebster) [inline kiva/fonttools/font.py:217]:
```suggestion
```

File: kiva/fonttools/font.py
Comment (jwiggins) [reply kiva/fonttools/font.py:196]: I think it's best to just not break existing code here and if we want to give a nicer interface for new code then that should be introduced.

File: enable/trait_defs/kiva_font_trait.py
Comment (corranwebster) [reply enable/trait_defs/kiva_font_trait.py:97]: The failure wasn't being discovered because the tests weren't in a package; that's been resolved and the tests should be working now.

There are a few other places where the tests aren't in packages - I suspect that this was a problem introduced when we moved from nose to unittest discover.

File: kiva/fonttools/font.py
Comment (corranwebster) [reply kiva/fonttools/font.py:197]: I agree.  But `RuntimeError` is what the code previously raised in this situation.  I'll open an issue.

Comment (corranwebster) [reply kiva/fonttools/font.py:197]: Opened #933

File: kiva/fonttools/font.py
Comment (corranwebster) [reply kiva/fonttools/font.py:196]: I've added #933 to track this.

File: kiva/trait_defs/tests/test_kiva_font_trait.py
Comment (corranwebster) [reply kiva/trait_defs/tests/test_kiva_font_trait.py:20]: Opened #934

**Comment by corranwebster** [comment]
I think that's everything addressed; assuming tests pass, this is ready for re-review.

File: enable/trait_defs/kiva_font_trait.py
Comment (mdickinson) [reply enable/trait_defs/kiva_font_trait.py:117]: This still needs addressing; I think Traits master will break this code if we don't set `default_value_type` as well as `default_value`.

File: enable/tests/trait_defs/test_kiva_font_trait.py
Comment (mdickinson) [inline enable/tests/trait_defs/test_kiva_font_trait.py]:
```suggestion
    KivaFont, font_styles, font_weights
```

File: enable/examples/demo/enable/editors/font_editor.py
Comment (mdickinson) [inline enable/examples/demo/enable/editors/font_editor.py]:
```suggestion
```

**Comment by mdickinson** [comment]
Thanks for the updates! There are a number of nitpick-level comments and suggestions left, but the only remaining comment that I think is important to address is this one: https://github.com/enthought/enable/pull/929/files#r860571004. If we don't set the default value type, then I think Traits master _will_ break this code.

**Comment by mdickinson** [comment]
Aargh; sorry. I was somehow seeing out-of-date code in GitHub. All looks good.

**Review by mdickinson — APPROVED** [review]
LGTM

File: enable/trait_defs/kiva_font_trait.py
Comment (mdickinson) [reply enable/trait_defs/kiva_font_trait.py:117]: Sorry; GitHub reading fail on my part. Looks good.

**Pull request merged**
